### DISCUSSION TITLE
Lazy task launcher for ktx-async

### DISCUSSION THREAD

**Discussion by PriNova** [libktx/ktx#420] [Ideas]
When I understand it correctly, that if I start an unconditional launch or async in the render loop, every frame a new job/deferred will be created. 
The usual use case will be an event to start launch/async. For example an UI event or timed event.

Now, I'm using background async's at every frame for heavy calculations and syncing them when completed. Yes, it could skip frames when needed longer.
I found out that at every frame a new deferred will be created and handling them for syncing let me involve some management. For this I created a helper map which stores the deferred with a given name and checkings for completion to update it.

I wrote a little .kt for it and maybe it fits into this project. I will attach it here and maybe it want to be incorporated after review.

The use case in the render-loop is this:

```kotlin
NamedLazyJob(name = "GenerateBoxes", dispatcher = dis) {
    repeat(30) {
	pos.add(Center + RandomPointInRange(-scale, scale))
    }
}.startIf {frame >= 100L}
.remove("GenerateBoxes")
```
It creates a named lazy launch and add's at every frame 30 random point-coodinates centered at the middle of the screen starting at frameId 100. After the named launch will be removed from the map.
In this example it creates every frame a launch but managed.
Another use-case could be to start a named lazy launch for a single event and so only one launch/async will be lazily created and executed.

What do you think about it? Or is this a edge case?

https://pastebin.com/611NsGPY

**Comment by czyzby** [comment]
So your goal is to skip a specific number of frames before running a task? Keep in mind that the FPS might vary for the users, so relying on delta time/specific time intervals is more consistent. Have you tried using `delay` or `schedule`?

I think this utility might be a bit too project-specific. The current combination of `launch`, `delay`, `skipFrame`, `schedule` and `interval` seems to cover the most common use cases when it comes to async operations and timed events. A similar effect can be achieved by launching a coroutine in the `render` method when a specific condition is met. It's cool that you've managed the extend the library for your needs though.

**Reply by PriNova** [reply]
Thank you for your answer.

You're right. This is more project-specific. At least it abstract things away and about the managed launches/asyncs, with a well defined event system, like the scheduler and timer, the jobmanager seems obsolete.

I find your extension really great and wished all libgdx will be written in kotlin :-)

P.S.: Maybe the jobmanager helps if deferrers needs longer than a frametick and finishes job after ie. 3 frames. Instead of creating every frame a new deferrer it needs to check, if the old one is completed and then a new one will be started
